File: sdk-skills/gjc-sdk-author/templates/direct-sdk.py
#!/usr/bin/env python3

from __future__ import annotations

# Trusted-local procedural policy only. The Broker and SessionRouter retain session lifecycle and attachment authority.

import argparse
import hashlib
import json
import re
import secrets
import subprocess
import sys
from typing import Any, NoReturn

# Long-running prompts: the SDK deadline is a progress-aware lease (sdk.promptDeadlineMs is
# an inactivity lease renewed only by attributable tool_execution_start/update/end for the exact
# accepted commandId/turnId, bounded by sdk.promptMaxRuntimeMs; a running tool's partial-result
# tool_execution_update keeps a long-running tool alive mid-run). Persist session_id/turn_id
# and reconcile via turn.result (Q26) rather than blindly replaying; heartbeats/assistant-text/thinking/
# retries/other-turn activity do not renew. Distinguish the bounded await_turn poll timeout
# from the SDK terminal deadline.

CORE_QUERIES = (
    "session.metadata",
    "context.get",
    "goal.list/get",
    "todo.list",
    "workflow.gates.list",
    "session.stats",
)
ALLOWED_CONTROLS = ['turn.prompt','turn.steer','turn.follow_up','ask.answer','workflow.gate_answer','todo.replace','session.switch','session.rename']
SECRET_FIELD = re.compile(r"(?:secret|token|password|credential|authorization|api[_-]?key)", re.IGNORECASE)


class SafeArgumentParser(argparse.ArgumentParser):
    def error(self, message: str) -> NoReturn:
        raise ValueError("invalid_argument")


def parse_args() -> argparse.Namespace:
    parser = SafeArgumentParser(description="Trusted local broker-bound GJC session template", allow_abbrev=False)
    parser.add_argument("--repo", required=True)
    parser.add_argument("--session-id", required=True)
    parser.add_argument("--mode", choices=("inspect", "control"), default="inspect")
    parser.add_argument("--operation")
    parser.add_argument("--input", default="{}")
    return parser.parse_args()


def has_secret_field(value: Any) -> bool:
    if isinstance(value, list):
        return any(has_secret_field(item) for item in value)
    if not isinstance(value, dict):
        return False
    return any(SECRET_FIELD.search(key) or has_secret_field(item) for key, item in value.items())


def redact(value: Any) -> Any:
    if isinstance(value, list):
        return [redact(item) for item in value]
    if isinstance(value, dict):
        return {
            key: "[REDACTED]" if SECRET_FIELD.search(key) else redact(item)
            for key, item in value.items()
        }
    return value


class CliFailure(Exception):
    def __init__(self, envelope: dict[str, Any], exit_code: int) -> None:
        super().__init__("broker_request_failed")
        self.envelope = envelope
        self.exit_code = exit_code


def run_gjc_session(repo: str, arguments: list[str]) -> dict[str, Any]:
    try:
        completed = subprocess.run(
            ["gjc", "sdk", "session", *arguments, "--json"],
            cwd=repo,
            stdin=subprocess.DEVNULL,
            stdout=subprocess.PIPE,
            stderr=subprocess.PIPE,
            text=True,
            check=False,
            timeout=30,
        )
    except (OSError, subprocess.SubprocessError):
        raise ValueError("broker_request_failed") from None
    try:
        response: Any = json.loads(completed.stdout)
    except json.JSONDecodeError:
        raise ValueError("invalid_cli_response") from None
    if not isinstance(response, dict):
        raise ValueError("invalid_cli_response")
    if completed.returncode != 0 or response.get("ok") is False:
        if len(completed.stdout.encode("utf-8")) > 8192 or response.get("schema") != "gjc.command-error" or response.get("version") != 1 or response.get("ok") is not False:
            raise ValueError("invalid_cli_response")
        raise CliFailure(response, 2 if completed.returncode == 2 else 1)
    return response


def inspect(repo: str, session_id: str) -> dict[str, Any]:
    snapshot: dict[str, Any] = {}
    for query in CORE_QUERIES:
        try:
            response = run_gjc_session(repo, ["raw", "query", session_id, "--query", query])
            if response.get("ok") is False:
                raise ValueError("query_unavailable")
            snapshot[query] = {"status": "confirmed", "source": query, "value": redact(response)}
        except CliFailure as error:
            snapshot[query] = {"status": "unavailable", "source": query, "failure": error.envelope}
        except Exception:
            snapshot[query] = {"status": "unavailable", "source": query}
    return snapshot


def require_approval(session_id: str, operation: str, operation_input: dict[str, Any]) -> None:
    payload = json.dumps(
        {"sessionId": session_id, "operation": operation, "input": operation_input},
        sort_keys=True,
        separators=(",", ":"),
    )
    digest = hashlib.sha256(payload.encode("utf-8")).hexdigest()[:16]
    challenge = f"APPROVE {session_id} {operation} {digest} {secrets.token_hex(8)}"
    print(f"Approval required: {challenge}\nType the exact challenge: ", file=sys.stderr, end="", flush=True)
    answer = sys.stdin.readline()
    if answer.strip() != challenge:
        raise ValueError("human_approval_required")


def main() -> None:
    args = parse_args()
    operation_input = json.loads(args.input)
    if not isinstance(operation_input, dict):
        raise ValueError("input must be an object")
    if has_secret_field(operation_input):
        raise ValueError("secret_input_forbidden")
    if args.mode == "inspect":
        result = inspect(args.repo, args.session_id)
        print(json.dumps({"sessionId": args.session_id, "result": result}, indent=2))
        return
    operation = args.operation
    if operation is None or operation not in ALLOWED_CONTROLS:
        raise ValueError("operation_not_allowed")
    if operation == "workflow.gate_answer":
        operation_input = {**operation_input, "expectedSessionId": args.session_id}
    require_approval(args.session_id, operation, operation_input)
    result = run_gjc_session(
        args.repo,
        [
            "raw",
            "control",
            args.session_id,
            "--op",
            operation,
            "--json-input",
            json.dumps(operation_input, separators=(",", ":")),
            "--confirm",
        ],
    )
    if result.get("ok") is False:
        raise ValueError("control_failed")
    print(json.dumps(redact({"sessionId": args.session_id, "result": result}), indent=2))


try:
    main()
except CliFailure as error:
    print(json.dumps(error.envelope, ensure_ascii=False, separators=(",", ":")))
    raise SystemExit(error.exit_code)
except Exception:
    print("GJC SDK request failed safely.", file=sys.stderr)
    raise SystemExit(1)

File: sdk-skills/gjc-sdk-author/templates/direct-sdk.ts
#!/usr/bin/env bun

import { createHash, randomBytes } from "node:crypto";
import { createInterface } from "node:readline/promises";

// Trusted-local procedural policy only. The Broker and SessionRouter retain session lifecycle and attachment authority.

// Long-running prompts: the SDK deadline is a progress-aware lease (sdk.promptDeadlineMs is
// an inactivity lease renewed only by attributable tool_execution_start/update/end for the exact
// accepted commandId/turnId, bounded by sdk.promptMaxRuntimeMs; a running tool's partial-result
// tool_execution_update keeps a long-running tool alive mid-run). Persist session_id/turn_id
// and reconcile via turn.result (Q26) rather than blindly replaying; heartbeats/assistant-text/thinking/
// retries/other-turn activity do not renew. Distinguish the bounded await_turn poll timeout
// from the SDK terminal deadline.

const CORE_QUERIES = [
	"session.metadata",
	"context.get",
	"goal.list/get",
	"todo.list",
	"workflow.gates.list",
	"session.stats",
] as const;

const ALLOWED_CONTROLS: ReadonlySet<string> = new Set(["turn.prompt","turn.steer","turn.follow_up","ask.answer","workflow.gate_answer","todo.replace","session.switch","session.rename"]);
const SECRET_FIELD = /(?:secret|token|password|credential|authorization|api[_-]?key)/i;
const ALLOWED_ARGUMENTS = new Set(["--repo", "--session-id", "--mode", "--operation", "--input"]);

type Arguments = {
	repo: string;
	sessionId: string;
	mode: "inspect" | "control";
	operation?: string;
	input: Record<string, unknown>;
};

function parseArgs(argv: string[]): Arguments {
	const values = new Map<string, string>();
	for (let index = 0; index < argv.length; index++) {
		const token = argv[index];
		if (!ALLOWED_ARGUMENTS.has(token) || values.has(token)) throw new Error("invalid_argument");
		const value = argv[++index];
		if (!value) throw new Error("missing_argument_value");
		values.set(token, value);
	}
	const repo = values.get("--repo");
	const sessionId = values.get("--session-id");
	if (!repo) throw new Error("missing_repo");
	if (!sessionId) throw new Error("missing_session_id");
	const mode = values.get("--mode") ?? "inspect";
	if (mode !== "inspect" && mode !== "control") throw new Error("invalid_mode");
	let input: Record<string, unknown> = {};
	const rawInput = values.get("--input");
	if (rawInput) {
		const parsed: unknown = JSON.parse(rawInput);
		if (!parsed || typeof parsed !== "object" || Array.isArray(parsed)) throw new Error("invalid_input");
		input = parsed as Record<string, unknown>;
	}
	if (hasSecretField(input)) throw new Error("secret_input_forbidden");
	return { repo, sessionId, mode, operation: values.get("--operation"), input };
}

function hasSecretField(value: unknown): boolean {
	if (Array.isArray(value)) return value.some(hasSecretField);
	if (!value || typeof value !== "object") return false;
	return Object.entries(value as Record<string, unknown>).some(([key, nested]) => SECRET_FIELD.test(key) || hasSecretField(nested));
}

function redact(value: unknown): unknown {
	if (Array.isArray(value)) return value.map(redact);
	if (!value || typeof value !== "object") return value;
	return Object.fromEntries(
		Object.entries(value as Record<string, unknown>).map(([key, item]) => [
			key,
			SECRET_FIELD.test(key) ? "[REDACTED]" : redact(item),
		]),
	);
}

function parseCliJson(stdout: string): Record<string, unknown> {
	let parsed: unknown;
	try {
		parsed = JSON.parse(stdout);
	} catch {
		throw new Error("invalid_cli_response");
	}
	if (!parsed || typeof parsed !== "object" || Array.isArray(parsed)) throw new Error("invalid_cli_response");
	return parsed as Record<string, unknown>;
}

// Preserve the trusted CLI's bounded public envelope, including exact reconciliation references.
class CliFailure extends Error {
	constructor(readonly envelope: Record<string, unknown>, readonly exitCode: number) { super("broker_request_failed"); }
}

async function runGjcSession(repo: string, arguments_: readonly string[]): Promise<Record<string, unknown>> {
	const child = Bun.spawn(["gjc", "sdk", "session", ...arguments_, "--json"], {
		cwd: repo,
		stdin: "ignore",
		stdout: "pipe",
		stderr: "pipe",
	});
	const [exitCode, stdout] = await Promise.all([
		child.exited,
		new Response(child.stdout).text(),
		new Response(child.stderr).text(),
	]);
	const response = parseCliJson(stdout);
	if (exitCode !== 0 || response.ok === false) {
		if (Buffer.byteLength(stdout) > 8192 || response.schema !== "gjc.command-error" || response.version !== 1 || response.ok !== false)
			throw new Error("invalid_cli_response");
		throw new CliFailure(response, exitCode === 2 ? 2 : 1);
	}
	return response;
}

async function inspect(repo: string, sessionId: string): Promise<Record<string, unknown>> {
	const snapshot: Record<string, unknown> = {};
	for (const query of CORE_QUERIES) {
		try {
			const response = await runGjcSession(repo, ["raw", "query", sessionId, "--query", query]);
			if (response.ok === false) throw new Error("query_unavailable");
			snapshot[query] = { status: "confirmed", source: query, value: redact(response) };
		} catch (error) {
			snapshot[query] = { status: "unavailable", source: query, ...(error instanceof CliFailure ? { failure: error.envelope } : {}) };
		}
	}
	return snapshot;
}

async function requireApproval(sessionId: string, operation: string, input: Record<string, unknown>): Promise<void> {
	const digest = createHash("sha256").update(JSON.stringify({ sessionId, operation, input })).digest("hex").slice(0, 16);
	const challenge = `APPROVE ${sessionId} ${operation} ${digest} ${randomBytes(8).toString("hex")}`;
	const reader = createInterface({ input: process.stdin, output: process.stderr });
	try {
		if ((await reader.question(`Approval required: ${challenge}\nType the exact challenge: `)).trim() !== challenge)
			throw new Error("human_approval_required");
	} finally { reader.close(); }
}


async function main(): Promise<void> {
	const args = parseArgs(process.argv.slice(2));
	if (args.mode === "inspect") {
		const result = await inspect(args.repo, args.sessionId);
		process.stdout.write(JSON.stringify({ sessionId: args.sessionId, result }, null, 2) + "\n");
		return;
	}
	if (!args.operation || !ALLOWED_CONTROLS.has(args.operation)) throw new Error("operation_not_allowed");
	const input = args.operation === "workflow.gate_answer" ? { ...args.input, expectedSessionId: args.sessionId } : args.input;
	await requireApproval(args.sessionId, args.operation, input);
	const result = await runGjcSession(args.repo, [
		"raw",
		"control",
		args.sessionId,
		"--op",
		args.operation,
		"--json-input",
		JSON.stringify(input),
		"--confirm",
	]);
	if (result.ok === false) throw new Error("control_failed");
	process.stdout.write(JSON.stringify(redact({ sessionId: args.sessionId, result }), null, 2) + "\n");
}

main().catch(error => {
	if (error instanceof CliFailure) {
		process.stdout.write(JSON.stringify(error.envelope) + "\n");
		process.exitCode = error.exitCode;
	} else { process.stderr.write("GJC SDK request failed safely.\n"); process.exitCode = 1; }
});
